url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/util/IOUtils.java

```diff
@@ -0,0 +1,289 @@
+	        	f.read(buffer);
+	        	str = new String(buffer);
+	        }
+	    } catch (FileNotFoundException e) {
+			e.printStackTrace();
+		} catch (IOException e) {
+			e.printStackTrace();
+		} finally {
+	        if (f != null) {
+	        	try { 
+	        		f.close(); 
+	        	} catch (IOException e) { 
+	        		e.printStackTrace();}
+	        	}
+	    }
+	    return str;
+	}
+	
+	public static NusmvSTS readNusmvSTSFromFile(File file) {
+		String str = IOUtils.readFileAsString(file);	
+		NusmvSTS newSTS = new NusmvSTS();
+	
+		for (String line : str.split("\n")) {
+			String[] split = line.split("->");
+			if (split.length == 2) {
+				String from = split[0];
+				Map<String, String> varMap = new HashMap<String, String>();
+				varMap.put("name.state", from);
+				State fromState = newSTS.addState(varMap);
+
+				String[] split2 = split[1].split("\\[label=\"");
+				if (split2.length == 2) {
+					String to = split2[0];
+					varMap.put("name.state", to);
+					State toState = newSTS.addState(varMap);
+
+					String[] acts = split2[1].split("\\\\n");
+					for (String act : acts) {
+						String[] split3 = act.split(" = ");
+
+						if (split3.length == 2) {
+							String var = split3[0];
+							String value = split3[1].substring(0, split3[1].length()-3);
+							if (value.endsWith("\"")) {
+								value = value.substring(0, value.length()-1);
+							}
+							varMap.clear();
+							varMap.put(var, value);
+
+							Action action = newSTS.addAction(varMap);
+							newSTS.addTransition(new NusmvTransition(fromState, action, toState));
+
+						}
+					}
+				}
+			}
+		}
+
+		return newSTS;
+	}
+	
+	public static STS readSTSFromFile(File file) {
+		String str = IOUtils.readFileAsString(file);	
+		STS newSTS = new DefaultSTS();
+	
+		for (String line : str.split("\n")) {
+			String[] split = line.split("->");
+			if (split.length == 2) {
+				String from = split[0];
+				State fromState = newSTS.getState(from);
+				if (fromState == null) {
+//					System.out.println("Creating state " + from);
+					fromState = new DefaultState(from);
+					newSTS.addState(fromState);
+				}
+
+				String[] split2 = split[1].split("\\[label=\"");
+				if (split2.length == 2) {
+					String to = split2[0];
+
+					State toState = newSTS.getState(to);
+					if (toState == null) {
+						toState = new DefaultState(to);
+						newSTS.addState(toState);
+					}
+					
+					String actName = split2[1].substring(0, split2[1].length() - 4);
+					if (actName.endsWith("\"")) {
+						actName = actName.substring(0, actName.length() - 1);
+					}
+					Action action = new DefaultAction(actName, true);
+					newSTS.addAction(action);
+					newSTS.addTransition(new NusmvTransition(fromState, action, toState));
+				}
+			}
+		}
+		
+		return newSTS;
+	}
+	
+	public static void writeStringToFile(String str, String filePath) {
+		File file = new File(filePath);
+		
+		try {
+			BufferedWriter out = new BufferedWriter(new FileWriter(file, false));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 146, 147. The resource is closed at line 148. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).